### ISSUE TITLE
cant install nixos because of src/libexpr/primops.cc:56: void nix::EvalState::realiseContext(const PathSet&): Assertion `store->isStorePath(ctx)' failed.

### ISSUE THREAD

**Issue by srghma** [NixOS/nixpkgs#42557]
### Steps to reproduce

Download nixos graphical iso

```

[root@nixos:~]# nix-shell -p nix-info --run "nix-info -m"
these paths will be fetched (0.32 MiB download, 1.90 MiB unpacked):
  /nix/store/5hszvi0hczzh25mmsz5rkb9qa758c10p-bash-4.4-p12-dev
  /nix/store/bkx6rzqciq0sdn1z5liard33ffvv1n3s-bash-4.4-p12-doc
copying path '/nix/store/bkx6rzqciq0sdn1z5liard33ffvv1n3s-bash-4.4-p12-doc' from 'https://cache.nixos.org'...
copying path '/nix/store/5hszvi0hczzh25mmsz5rkb9qa758c10p-bash-4.4-p12-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.51, NixOS, 18.03.132750.91b286c8935 (Impala)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.0.4`
 - channels(root): `"nixos-18.03.132750.91b286c8935"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`


[root@nixos:~]# cat /mnt/etc/nixos/configuration.nix 
import /mnt/home/srghma/.dotfiles/nixos/root/default.nix


#### it includes this file https://github.com/srghma/dotfiles/blob/master/nixos/root/default.nix

[root@nixos:~]# lsblk
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
loop0    7:0    0   977M  1 loop /nix/.ro-store
sda      8:0    0 931.5G  0 disk 
├─sda1   8:1    0   499M  0 part 
├─sda2   8:2    0   100M  0 part 
├─sda3   8:3    0    16M  0 part 
├─sda4   8:4    0 100.9G  0 part 
└─sda5   8:5    0 830.1G  0 part 
sdb      8:16   0 119.2G  0 disk 
├─sdb1   8:17   0   100M  0 part /mnt/boot
├─sdb2   8:18   0     8G  0 part 
└─sdb3   8:19   0 111.1G  0 part /mnt
sdc      8:32   1  28.9G  0 disk 
├─sdc1   8:33   1  1011M  0 part /iso
└─sdc2   8:34   1    16M  0 part 

[root@nixos:~]# nixos-install --no-bootloader --no-root-passwd --no-channel-copy
building the configuration in /mnt/etc/nixos/configuration.nix...
[13.1/13.1 MiB DL] downloading 'https://github.com/nixos/nixpkgs-channels/arcnix: src/libexpr/primops.cc:56: void nix::EvalState::realiseContext(const PathSet&): Assertion `store->isStorePath(ctx)' failed.
/run/current-system/sw/bin/nixos-install: line 98:  1665 Aborted                 nix build --out-link "$outLink" --store "$mountPoint" "${extraBuildFlags[@]}" --extra-substituters "$sub" -f '<nixpkgs/nixos>' system -I "nixos-config=$NIXOS_CONFIG"
```

**Comment by srghma** [comment]
though it installed fine 

```
[root@nixos:~]# mv /mnt/etc/nixos/configuration.nix /mnt/etc/nixos/configuration.nix__

[root@nixos:~]# nixos-generate-config --root /mnt
writing /mnt/etc/nixos/hardware-configuration.nix...
writing /mnt/etc/nixos/configuration.nix...

[root@nixos:~]# vim /mnt/etc/nixos/configuration.nix

[root@nixos:~]# nixos-install
building the configuration in /mnt/etc/nixos/configuration.nix...
[71 built, 16 copied (35.8 MiB), 5.3 MiB DL]
copying channel...
installing the boot loader...
removing group ‘docker’
removing group ‘networkmanager’
removing group ‘nm-openvpn’
removing group ‘lightdm’
removing user ‘nm-openvpn’
removing user ‘lightdm’
removing user ‘srghma’
removing user ‘rtkit’
removing user ‘nm-iodine’
removing user ‘sshd’
setting up /etc...
removing obsolete symlink ‘/etc/zinputrc’...
removing obsolete symlink ‘/etc/zprofile’...
removing obsolete symlink ‘/etc/zshenv’...
removing obsolete symlink ‘/etc/asound.conf’...
removing obsolete symlink ‘/etc/libao.conf’...
removing obsolete symlink ‘/etc/zshrc’...
removing obsolete symlink ‘/etc/bluetooth/main.conf’...
removing obsolete symlink ‘/etc/pulse/default.pa’...
removing obsolete symlink ‘/etc/pulse/daemon.conf’...
removing obsolete symlink ‘/etc/ssh/moduli’...
removing obsolete symlink ‘/etc/openal/alsoft.conf’...
removing obsolete symlink ‘/etc/NetworkManager/NetworkManager.conf’...
removing obsolete symlink ‘/etc/NetworkManager/dispatcher.d/02overridedns’...
removing obsolete symlink ‘/etc/NetworkManager/VPN/nm-openvpn-service.name’...
removing obsolete symlink ‘/etc/NetworkManager/VPN/nm-l2tp-service.name’...
removing obsolete symlink ‘/etc/NetworkManager/VPN/nm-vpnc-service.name’...
removing obsolete symlink ‘/etc/NetworkManager/VPN/nm-fortisslvpn-service.name’...
removing obsolete symlink ‘/etc/NetworkManager/VPN/nm-openconnect-service.name’...
removing obsolete symlink ‘/etc/NetworkManager/VPN/nm-strongswan-service.name’...
removing obsolete symlink ‘/etc/NetworkManager/VPN/nm-iodine-service.name’...
removing obsolete symlink ‘/etc/lightdm/lightdm.conf’...
removing obsolete symlink ‘/etc/lightdm/users.conf’...
removing obsolete symlink ‘/etc/lightdm/lightdm-gtk-greeter.conf’...
removing obsolete symlink ‘/etc/PackageKit/Vendor.conf’...
removing obsolete symlink ‘/etc/PackageKit/PackageKit.conf’...
removing obsolete symlink ‘/etc/X11/xorg.conf.d/40-libinput.conf’...
removing obsolete symlink ‘/etc/X11/xorg.conf.d/10-evdev.conf’...
removing obsolete symlink ‘/etc/X11/xorg.conf.d/00-keyboard.conf’...
removing obsolete symlink ‘/etc/pam.d/lightdm-greeter’...
removing obsolete symlink ‘/etc/pam.d/lightdm’...
removing obsolete symlink ‘/etc/pam.d/lightdm-autologin’...
removing obsolete symlink ‘/etc/pam.d/sshd’...
removing obsolete file ‘/etc/localtime’...
Copied "/nix/store/si8adck21xj42dm1h5k7zf3x6x9dv5fg-systemd-237/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/systemd/systemd-bootx64.efi".
Copied "/nix/store/si8adck21xj42dm1h5k7zf3x6x9dv5fg-systemd-237/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/BOOT/BOOTX64.EFI".
Created EFI boot entry "Linux Boot Manager".
setting up /etc...
setting root password...
New password: 
Retype new password: 
passwd: password updated successfully
installation finished!
```

**Comment by srghma** [comment]
I have stumbled on this error once again

I have found that the error

```
[root@nixos:~]# nixos-install
building the configuration in /mnt/etc/nixos/configuration.nix...
GC Warning: Bad initial heap size 100000 - ignoring it.
querying info about missing pathsnix: src/libexpr/primops.cc:56: void nix::EvalState::realiseContext(const PathSet&): Assertion `store->isStorePath(ctx)' failed.
/run/current-system/sw/bin/nixos-install: line 98:  9843 Aborted                 nix build --out-link "$outLink" --store "$mountPoint" "${extraBuildFlags[@]}" --extra-substituters "$sub" -f '<nixpkgs/nixos>' system -I "nixos-config=$NIXOS_CONFIG"

```

 disappears when I comment out this lines

https://github.com/srghma/dotfiles/blob/93b1f5820719dedc319724f83d37594210d35c8f/nixos/root/default.nix#L46-L49

**Comment by mickours** [comment]
I have the same issue with an assert at the same line.

It seems that the files imported with fetchGit are not evaluated properly. I'm not sure how to investigate this issue because the error message lacks information...

Note that the install works with a basic configuration.

> I have stumbled on this error once again
> 
> I have found that the error
> 
> ```
> [root@nixos:~]# nixos-install
> building the configuration in /mnt/etc/nixos/configuration.nix...
> GC Warning: Bad initial heap size 100000 - ignoring it.
> querying info about missing pathsnix: src/libexpr/primops.cc:56: void nix::EvalState::realiseContext(const PathSet&): Assertion `store->isStorePath(ctx)' failed.
> /run/current-system/sw/bin/nixos-install: line 98:  9843 Aborted                 nix build --out-link "$outLink" --store "$mountPoint" "${extraBuildFlags[@]}" --extra-substituters "$sub" -f '<nixpkgs/nixos>' system -I "nixos-config=$NIXOS_CONFIG"
> ```
> 
> disappears when I comment out this lines
> 
> https://github.com/srghma/dotfiles/blob/93b1f5820719dedc319724f83d37594210d35c8f/nixos/root/default.nix#L46-L49

**Comment by mickours** [comment]
Maybe this is related to the restricted evaluation mode:
https://nixos.org/nix/manual/#conf-restrict-eval
```
restrict-eval

    If set to true, the Nix evaluator will not allow access to any files outside of the Nix search path (as set via the NIX_PATH environment variable or the -I option), or to URIs outside of allowed-uri. The default is false.
```
Does anybody know if nix-install use this mode?

**Comment by joepie91** [comment]
Running into the same issue here, while trying to install 19.03 via `nixos-install` from a live 18.09 system (on the same physical box, but on a different disk):

```
[root@desktop-home:/mnt-new/etc/nixos]# nixos-install --no-bootloader --root /mnt-new
building the configuration in /mnt-new/etc/nixos/configuration.nix...
[0.0/0.0 MiB DL] downloading 'https://github.com/mozilla/nixpkgs-mozilla/archive/master.tar.gz'nix: src/libexpr/primops.cc:56: void nix::EvalState::realiseContext(const PathSet&): Assertion `store->isStorePath(ctx)' failed.
/run/current-system/sw/bin/nixos-install: line 98: 14197 Aborted                 nix build --out-link "$outLink" --store "$mountPoint" "${extraBuildFlags[@]}" --extra-substituters "$sub" -f '<nixpkgs/nixos>' system -I "nixos-config=$NIXOS_CONFIG"
```

I have a custom root, `/mnt-new`, because this is being installed from a live system where `/mnt` is already in use for storing various other mountpoints.

It's unclear to me how to proceed.

**Comment by joepie91** [comment]
Probably a duplicate of NixOS/nix#2405.

**Comment by chrissound** [comment]
Also running into this issue... :/ Did any of you manage to find a solution to this?

**Comment by jamescostian** [comment]
Same issue here using the NixOS 19.09 installer - specifically `nixos-graphical-19.09.1690.0d9055a2ac2-x86_64-linux.iso`

Removing the `builtins.fetchTarball` and the stuff that used that tarball from my config makes the whole thing install, no errors

**Comment by hyperfekt** [comment]
My understanding is that this was fixed with Nix 2.3.1 via https://github.com/NixOS/nix/pull/3112.
The issue can be closed.

**Comment by evils** [comment]
closing on the assumption @hyperfekt knows best
last report here is for a 19.09 image that still had nix 2.3
